exonum_derive/src/tx_set.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
+use syn::{Data, DataEnum, DeriveInput, Type};
+
+struct TxSetVariant(u16, Ident, Type);
+
+fn get_tx_variants(data: &DataEnum) -> Vec<TxSetVariant> {
+    if data.variants.is_empty() {
+        panic!("TransactionSet enum should not be empty");
+    }
+
+    data.variants
+        .iter()
+        .enumerate()
+        .map(|(n, v)| {
+            if v.fields.iter().len() > 1 {
+                panic!("TransactionSet enum variant should have one field inside.");
+            }
+            let field = v
+                .fields
+                .iter()
+                .next()
+                .expect("TransactionSet enum variant can't be empty.");
+            TxSetVariant(n as u16, v.ident.clone(), field.ty.clone())
+        }).collect()
+}
+
+fn gen_conversions_for_transactions(
+    name: &Ident,
+    variants: &[TxSetVariant],
+    cr: &quote::ToTokens,
+) -> impl quote::ToTokens {
+    let conversions = variants.iter().map(|TxSetVariant(_, id, ty)| {
+        quote! {
+          impl Into<#name> for #ty {
+               fn into(self) -> #name {
+                     #name::#id(self)
+               }
+          }
+
+          impl Into<#cr::messages::ServiceTransaction> for #ty {
+              fn into(self) -> #cr::messages::ServiceTransaction {
+                  let set: #name = self.into();
+                  set.into()
+              }
+          }
+        }
+    });
+
+    quote!{
+        #(#conversions)*
+    }
+}
+
+fn gen_into_service_tx(
+    name: &Ident,
+    variants: &[TxSetVariant],
+    cr: &quote::ToTokens,
+) -> impl quote::ToTokens {
+    let tx_set_impl = variants.iter().map(|TxSetVariant(n, id, _)| {
+        quote! {
+            #name::#id( ref tx) => ( #n, tx.encode().unwrap()),
```

Comment:
Is this `encode()` from `BinaryForm`? I'm slightly concerned about the complexity of all these traits...

Reply:
Yes it is from `BinaryForm`.
What complexity are you talking about?
If you talk about serialization than there is only one important trait: `BinaryForm`(there is also `StorageValue` which is `BinaryForm` without fail).
There is also `ProtobufConvert` to make validation of protobuf structs simpler.
Or you talking about Transaction related traits?